File: engine/compass/documents.py
"""Document storage for Compass workspaces.

Documents are stored as individual JSON files in .compass/documents/.
"""

from __future__ import annotations

import json
import logging
import uuid
from datetime import datetime
from pathlib import Path

from compass.models.documents import StoredDocument

logger = logging.getLogger(__name__)


def _docs_dir(workspace_path: Path) -> Path:
    """Get the documents directory for a workspace."""
    d = workspace_path / ".compass" / "documents"
    d.mkdir(parents=True, exist_ok=True)
    return d


def save_document(workspace_path: Path, doc: StoredDocument) -> StoredDocument:
    """Save a document to disk."""
    doc.updated_at = datetime.now().isoformat()
    doc_path = _docs_dir(workspace_path) / f"{doc.id}.json"
    doc_path.write_text(json.dumps(doc.model_dump(), indent=2))
    logger.info("Saved document %s: %s", doc.id, doc.title)
    return doc


def get_document(workspace_path: Path, doc_id: str) -> StoredDocument | None:
    """Get a document by ID."""
    doc_path = _docs_dir(workspace_path) / f"{doc_id}.json"
    if not doc_path.exists():
        return None
    try:
        data = json.loads(doc_path.read_text())
        return StoredDocument(**data)
    except (json.JSONDecodeError, Exception) as e:
        logger.warning("Failed to load document %s: %s", doc_id, e)
        return None


def list_documents(workspace_path: Path) -> list[StoredDocument]:
    """List all documents in a workspace."""
    docs_dir = _docs_dir(workspace_path)
    docs = []
    for f in sorted(docs_dir.glob("*.json"), key=lambda p: p.stat().st_mtime, reverse=True):
        try:
            data = json.loads(f.read_text())
            docs.append(StoredDocument(**data))
        except (json.JSONDecodeError, Exception):
            continue
    return docs


def delete_document(workspace_path: Path, doc_id: str) -> bool:
    """Delete a document by ID."""
    doc_path = _docs_dir(workspace_path) / f"{doc_id}.json"
    if doc_path.exists():
        doc_path.unlink()
        logger.info("Deleted document %s", doc_id)
        return True
    return False


def create_document(
    workspace_path: Path,
    title: str,
    doc_type: str = "custom",
    content_json: dict | None = None,
    content_markdown: str = "",
    tags: list[str] | None = None,
    evidence_ids: list[str] | None = None,
) -> StoredDocument:
    """Create a new document."""
    doc = StoredDocument(
        id=str(uuid.uuid4())[:8],
        title=title,
        doc_type=doc_type,
        content_json=content_json or {},
        content_markdown=content_markdown,
        tags=tags or [],
        evidence_ids=evidence_ids or [],
    )
    return save_document(workspace_path, doc)

File: app/src/stores/documents.ts
import { create } from "zustand";

export interface StoredDocument {
  id: string;
  title: string;
  doc_type: string;
  content_json: Record<string, unknown>;
  content_markdown: string;
  created_at: string;
  updated_at: string;
  tags: string[];
  evidence_ids: string[];
}

interface DocumentsState {
  documents: StoredDocument[];
  activeDoc: StoredDocument | null;
  loading: boolean;
  error: string | null;

  loadDocuments: (workspacePath: string) => Promise<void>;
  loadDocument: (workspacePath: string, docId: string) => Promise<void>;
  saveDocument: (workspacePath: string, doc: StoredDocument) => Promise<void>;
  createDocument: (
    workspacePath: string,
    title: string,
    docType: string,
  ) => Promise<StoredDocument | null>;
  deleteDocument: (workspacePath: string, docId: string) => Promise<void>;
  setActiveDoc: (doc: StoredDocument | null) => void;
}

export const useDocumentsStore = create<DocumentsState>((set, get) => ({
  documents: [],
  activeDoc: null,
  loading: false,
  error: null,

  loadDocuments: async (workspacePath) => {
    set({ loading: true, error: null });
    try {
      const res = (await window.compass.engine.call("/documents/list", {
        workspace_path: workspacePath,
      })) as { status: string; documents: StoredDocument[] };
      if (res.status === "ok") {
        set({ documents: res.documents, loading: false });
      } else {
        set({ error: "Failed to load documents", loading: false });
      }
    } catch (err) {
      set({
        error: err instanceof Error ? err.message : "Failed to load documents",
        loading: false,
      });
    }
  },

  loadDocument: async (workspacePath, docId) => {
    try {
      const res = (await window.compass.engine.call("/documents/get", {
        workspace_path: workspacePath,
        id: docId,
      })) as { status: string; document: StoredDocument };
      if (res.status === "ok") {
        set({ activeDoc: res.document });
      }
    } catch {
      // ignore
    }
  },

  saveDocument: async (workspacePath, doc) => {
    try {
      const res = (await window.compass.engine.call("/documents/save", {
        workspace_path: workspacePath,
        id: doc.id,
        title: doc.title,
        doc_type: doc.doc_type,
        content_json: doc.content_json,
        content_markdown: doc.content_markdown,
        tags: doc.tags,
        evidence_ids: doc.evidence_ids,
      })) as { status: string; document: StoredDocument };
      if (res.status === "ok") {
        set({ activeDoc: res.document });
        get().loadDocuments(workspacePath);
      }
    } catch {
      // ignore
    }
  },

  createDocument: async (workspacePath, title, docType) => {
    try {
      const res = (await window.compass.engine.call("/documents/save", {
        workspace_path: workspacePath,
        title,
        doc_type: docType,
      })) as { status: string; document: StoredDocument };
      if (res.status === "ok") {
        get().loadDocuments(workspacePath);
        return res.document;
      }
      return null;
    } catch {
      return null;
    }
  },

  deleteDocument: async (workspacePath, docId) => {
    try {
      await window.compass.engine.call("/documents/delete", {
        workspace_path: workspacePath,
        id: docId,
      });
      set((s) => ({
        documents: s.documents.filter((d) => d.id !== docId),
        activeDoc: s.activeDoc?.id === docId ? null : s.activeDoc,
      }));
    } catch {
      // ignore
    }
  },

  setActiveDoc: (doc) => set({ activeDoc: doc }),
}));
